src/prof_recent.c

```diff
@@ -506,33 +519,106 @@ prof_recent_alloc_dump(tsd_t *tsd, void (*write_cb)(void *, const char *),
 	emitter_init(&emitter, emitter_output_json_compact,
 	    buf_writer_get_write_cb(&buf_writer),
 	    buf_writer_get_cbopaque(&buf_writer));
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_dump_mtx);
 	emitter_begin(&emitter);
 
 	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
 	prof_recent_alloc_assert_count(tsd);
+	prof_recent_t *dump_dummy = prof_recent_alloc_dummy;
+	assert(dump_dummy != NULL);
+	prof_recent_alloc_dummy = prof_recent_alloc_dummy_reserve;
+	assert(prof_recent_alloc_dummy != NULL);
+	assert(prof_recent_alloc_dummy->next == prof_recent_alloc_dummy);
+	const ssize_t dump_count = prof_recent_alloc_count;
+	prof_recent_alloc_count = 0;
+	const ssize_t dump_max = prof_recent_alloc_max_get(tsd);
+	prof_recent_alloc_assert_count(tsd);
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
 
-	/*
-	 * Set prof_recent_alloc_max to 0 so that dumping won't block sampled
-	 * allocations: the allocations can complete but will not be recorded.
-	 */
-	ssize_t max = prof_recent_alloc_max_update(tsd, 0);
+	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize,
+		&dump_max);
+	emitter_json_array_kv_begin(&emitter, "recent_alloc");
 
-	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize, &max);
+	buf_writer_set_manual(&buf_writer);
+	prof_recent_t *node = NULL;
+	prof_recent_t *last = NULL;
+	while (true) {
+		malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+		if (node == NULL) {
+			node = dump_dummy->next;
```

Comment:
I put it here mainly because I wanted it to be under the `prof_recent_alloc_mtx` lock. Strictly speaking, this `dump_dummy->next` doesn't have to be under the lock, because we know that `dump_dummy` points to a list where only the content of the nodes rather than the nodes' linkages might be changed. But I'm just putting it under the lock to be rigorous, and to also be safe in case we might change any logic here in the future.

Reply:
Sounds good. Reading under the lock is the right thing to do.